File: codes/javascript/claude/complete/task_2/Solution.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

function reallocateBed(patient, hospitals) {
  const isHighRisk = patient.is_post_surgery;
  let result = {
    patient_id: patient.id,
    assigned_hospital: null,
    assigned_ward: null,
    assigned_bed: null,
    reallocated_patients: []
  };

  // Direct Assignment
  for (const hospital of hospitals) {
    if (isHighRisk && hospital.model !== 4) continue;
    if (!isHighRisk && hospital.model === 4) continue;

    for (const ward of hospital.wards) {
      if (isWardSuitable(patient, ward)) {
        const availableBed = ward.available_beds[0];
        if (availableBed) {
          result.assigned_hospital = hospital.id;
          result.assigned_ward = ward.ward_name;
          result.assigned_bed = availableBed;
          return result;
        }
      }
    }
  }

  // Reallocation for high-risk patients
  if (isHighRisk) {
    for (const hospital of hospitals.filter(h => h.model === 4)) {
      for (const ward of hospital.wards) {
        if (isWardSuitable(patient, ward)) {
          const transferablePatient = findTransferablePatient(ward, hospitals);
          if (transferablePatient) {
            result.assigned_hospital = hospital.id;
            result.assigned_ward = ward.ward_name;
            result.assigned_bed = transferablePatient.bed_number;
            result.reallocated_patients.push({
              patient_id: transferablePatient.id,
              new_hospital: transferablePatient.new_hospital,
              new_ward: transferablePatient.new_ward,
              new_bed: transferablePatient.new_bed
            });
            return result;
          }
        }
      }
    }
  }

  return result;
}

function isWardSuitable(patient, ward) {
  const genderMatch = ward.gender_restriction === 'No Restriction' ||
    (patient.gender === 'Male' && ward.gender_restriction === 'Male Only') ||
    (patient.gender === 'Female' && ward.gender_restriction === 'Female Only');
  
  const ageMatch = patient.age >= ward.age_restriction[0] && patient.age <= ward.age_restriction[1];
  
  const requirementsMatch = (patient.special_requirements || []).every(req => 
    ward.special_requirements.includes(req)
  );

  return genderMatch && ageMatch && requirementsMatch;
}

function findTransferablePatient(ward, hospitals) {
  const transferableCandidates = ward.current_patients
    .filter(p => p.days_in_hospital > 3 && !p.non_transferable)
    .sort((a, b) => b.days_in_hospital - a.days_in_hospital || a.bed_number - b.bed_number);

  for (const patient of transferableCandidates) {
    for (const hospital of hospitals.filter(h => h.model < 4)) {
      for (const newWard of hospital.wards) {
        if (isWardSuitable(patient, newWard) && newWard.available_beds.length > 0) {
          return {
            ...patient,
            new_hospital: hospital.id,
            new_ward: newWard.ward_name,
            new_bed: newWard.available_beds[0]
          };
        }
      }
    }
  }

  return null;
}

rl.on('line', (input) => {
  const { patient, hospitals } = JSON.parse(input);
  const result = reallocateBed(patient, hospitals);
  console.log(JSON.stringify(result));
  rl.close();
});

File: codes/python/claude/complete/task_2/Solution.py
import json
import sys

def reallocate_bed(patient, hospitals):
    def is_ward_suitable(ward, patient):
        gender_match = ward['gender_restriction'] == 'No Restriction' or \
                       (ward['gender_restriction'] == 'Male Only' and patient['gender'] == 'Male') or \
                       (ward['gender_restriction'] == 'Female Only' and patient['gender'] == 'Female')
        age_match = ward['age_restriction'][0] <= patient['age'] <= ward['age_restriction'][1]
        requirements_match = all(req in ward['special_requirements'] for req in patient['special_requirements'])
        return gender_match and age_match and requirements_match

    def find_direct_assignment(patient, hospitals):
        for hospital in hospitals:
            if patient['is_post_surgery'] and hospital['model'] != 4:
                continue
            if not patient['is_post_surgery'] and hospital['model'] == 4:
                continue
            for ward in hospital['wards']:
                if is_ward_suitable(ward, patient) and ward['available_beds']:
                    return hospital['id'], ward['ward_name'], min(ward['available_beds'])
        return None, None, None

    def find_reallocation(patient, hospitals):
        for hospital in hospitals:
            if hospital['model'] != 4:
                continue
            for ward in hospital['wards']:
                if not is_ward_suitable(ward, patient):
                    continue
                transferable_patients = sorted(
                    [p for p in ward['current_patients'] if p['days_in_hospital'] > 3 and not p['non_transferable']],
                    key=lambda x: (-x['days_in_hospital'], x['bed_number'])
                )
                for tp in transferable_patients:
                    new_assignment = find_direct_assignment(tp, [h for h in hospitals if h['model'] < 4])
                    if new_assignment[0]:
                        return hospital['id'], ward['ward_name'], tp['bed_number'], tp, new_assignment
        return None, None, None, None, None

    assigned_hospital, assigned_ward, assigned_bed = find_direct_assignment(patient, hospitals)
    reallocated_patients = []

    if not assigned_hospital and patient['is_post_surgery']:
        assigned_hospital, assigned_ward, assigned_bed, transferable_patient, new_assignment = find_reallocation(patient, hospitals)
        if assigned_hospital:
            reallocated_patients.append({
                "patient_id": transferable_patient['id'],
                "new_hospital": new_assignment[0],
                "new_ward": new_assignment[1],
                "new_bed": new_assignment[2]
            })

    return {
        "patient_id": patient['id'],
        "assigned_hospital": assigned_hospital,
        "assigned_ward": assigned_ward,
        "assigned_bed": assigned_bed,
        "reallocated_patients": reallocated_patients
    }

def main():
    input_data = json.loads(sys.stdin.read())
    result = reallocate_bed(input_data['patient'], input_data['hospitals'])
    print(json.dumps(result))

if __name__ == "__main__":
    main()